hmac.go

```diff
@@ -67,50 +77,84 @@ type opensslHMAC struct {
 	sum       []byte
 }
 
-func newHMAC1(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
+func newHMAC1(key []byte, md C.GO_EVP_MD_PTR) hmacCtx1 {
 	ctx := hmacCtxNew()
 	if ctx == nil {
 		panic("openssl: EVP_MAC_CTX_new failed")
 	}
 	if C.go_openssl_HMAC_Init_ex(ctx, unsafe.Pointer(&key[0]), C.int(len(key)), md, nil) == 0 {
 		panic(newOpenSSLError("HMAC_Init_ex failed"))
 	}
-	hmac := &opensslHMAC{
-		size:      h.Size(),
-		blockSize: h.BlockSize(),
-		ctx1:      hmacCtx1{ctx},
-	}
-	runtime.SetFinalizer(hmac, (*opensslHMAC).finalize)
-	return hmac
+	return hmacCtx1{ctx}
 }
 
-func newHMAC3(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
-	fetchHMACOnce.Do(func() {
-		name := C.CString("HMAC")
-		evpHMAC = C.go_openssl_EVP_MAC_fetch(nil, name, nil)
-		C.free(unsafe.Pointer(name))
-	})
-	if evpHMAC == nil {
+var hmacDigestsSupported sync.Map
+var fetchHMAC3 = sync.OnceValue(func() C.GO_EVP_MAC_PTR {
+	name := C.CString("HMAC")
+	mac := C.go_openssl_EVP_MAC_fetch(nil, name, nil)
+	C.free(unsafe.Pointer(name))
+	if mac == nil {
 		panic("openssl: HMAC not supported")
 	}
-	ctx := C.go_openssl_EVP_MAC_CTX_new(evpHMAC)
-	if ctx == nil {
-		panic("openssl: EVP_MAC_CTX_new failed")
-	}
-	digest := C.go_openssl_EVP_MD_get0_name(md)
+	return mac
+})
+
+func buildHMAC3Params(digest *C.char) C.GO_OSSL_PARAM_PTR {
 	bld := C.go_openssl_OSSL_PARAM_BLD_new()
 	if bld == nil {
 		panic(newOpenSSLError("OSSL_PARAM_BLD_new"))
 	}
 	defer C.go_openssl_OSSL_PARAM_BLD_free(bld)
 	C.go_openssl_OSSL_PARAM_BLD_push_utf8_string(bld, paramDigest, digest, 0)
-	params := C.go_openssl_OSSL_PARAM_BLD_to_param(bld)
+	return C.go_openssl_OSSL_PARAM_BLD_to_param(bld)
+}
+
+func isHMAC3DigestSupported(digest string) bool {
+	if _, ok := hmacDigestsSupported.Load(digest); ok {
+		return true
```

Comment:
Should return the value:

```suggestion
	if v, ok := hmacDigestsSupported.Load(digest); ok {
		return v.(bool)
```

I made this change to `TestHMAC` to try it out in `mcr.microsoft.com/oss/go/microsoft/golang:1.23-azurelinux3.0`:

```go
			h := NewHMAC(tt.fn, nil)
			if h == nil {
				h = NewHMAC(tt.fn, nil)
				if h == nil {
					t.Skip("not supported")
				}
			}
```

Reply:
Wops, good catch.